mrjob/dataproc.py

```diff
@@ -0,0 +1,873 @@
+        assert step['type'] == 'streaming', 'Jar not implemented'
+        main_jar_uri = _HADOOP_STREAMING_JAR_URI
+
+
+        # TODO - mtai @ davidmarin - Not clear if we should move _upload_args to file_uris, currently works fine as-is
+        args.extend(self._upload_args(self._upload_mgr))
+
+        # TODO - mtai @ davidmarin - Not clear if we should move some of these to properties dict to file_uris, currently works fine as-is
+        args.extend(self._hadoop_args_for_step(step_num))
+
+        mapper, combiner, reducer = (self._hadoop_streaming_commands(step_num))
+
+        if mapper:
+            args += ['-mapper', mapper]
+
+        if combiner:
+            args += ['-combiner', combiner]
+
+        if reducer:
+            args += ['-reducer', reducer]
+
+        for current_input_uri in self._step_input_uris(step_num):
+            args += ['-input', current_input_uri]
+
+        args += ['-output', self._step_output_uri(step_num)]
+
+        # TODO - mtai @ davidmarin - Add back support to specify a different jar URI
+        output_hadoop_job = {
+            "args": args,
+            "fileUris": file_uris,
+            "archiveUris": archive_uris,
+            "properties": properties,
+            "mainJarFileUri": main_jar_uri
+        }
+        return output_hadoop_job
+
+    def _launch_dataproc_cluster(self):
+        """Create an empty cluster on Dataproc, and set self._gcp_cluster to
+        its ID."""
+        self._wait_for_gcs_eventual_consistency()
+
+        self._gcp_cluster_start = time.time()
+
+        # make sure we can see the files we copied to GCS
+        log.info('Creating Dataproc Hadoop cluster')
+
+        # NOTE - mtai @ davidmarin - We create the cluster if its missing, otherwise we join an existing one
+        # NOTE - mtai @ davidmarin - The cluster MUST be created before submitting jobs, otherwise job submission fails
+
+        # TODO - mtai @ davidmarin - change the way we auto-create cluster names
+        # "clusterName must be a match of regex '(?:[a-z](?:[-a-z0-9]{0,53}[a-z0-9])?).'"
+        self._gcp_cluster = self._gcp_cluster or "default-%s" % self._opts['owner']
+
+        try:
+            self._get_cluster()
+            log.info('Adding our job to existing cluster %s' % self._gcp_cluster)
+        # TODO - mtai @ davidmarin - Change Exception to HttpError and check 404 for cluster does not exist
+        except Exception as e:
+            self._create_cluster()
+            log.info('Created new cluster %s' % self._gcp_cluster)
+
+        # keep track of when we launched our job
+        self._dataproc_job_start = time.time()
+
+    def _wait_for_steps_to_complete(self):
+        """Wait for every step of the job to complete, one by one."""
+
+        total_steps = self._num_steps()
+        # define out steps
+        for step_num in xrange(total_steps):
+            # NOTE - mtai @ davidmarin - Dataproc accepts multiple jobs at once
```

Comment:
Yeah, these names aren't really appropriate for Dataproc's model. :) You probably want methods named `_run_steps()`, `_launch_step()`, and `_wait_for_step_to_complete()`, or something like that.